src/libstore/filetransfer.cc

```diff
@@ -798,36 +870,55 @@ struct curlFileTransfer : public FileTransfer
+        httpsUrl.path = "/" + parsed.bucket + "/" + parsed.key;
+
+        // Handle endpoint configuration
+        if (auto endpoint = parsed.endpoint; std::holds_alternative<ParsedURL>(endpoint)) {
+            // Endpoint is already a ParsedURL (e.g., http://server:9000)
+            auto endpointUrl = std::get<ParsedURL>(endpoint);
+            httpsUrl.scheme = endpointUrl.scheme;
+            httpsUrl.authority = endpointUrl.authority;
+            // Append bucket and key to any existing path
+            if (!endpointUrl.path.empty() && endpointUrl.path != "/") {
+                httpsUrl.path = endpointUrl.path + "/" + parsed.bucket + "/" + parsed.key;
+            }
+        } else if (std::holds_alternative<ParsedURL::Authority>(endpoint)) {
+            // Endpoint is just an authority (hostname/port)
+            httpsUrl.scheme = scheme;
+            httpsUrl.authority = std::get<ParsedURL::Authority>(endpoint);
+        } else {
+            // No custom endpoint, use standard AWS S3 endpoint
+            httpsUrl.scheme = scheme;
+            httpsUrl.authority = ParsedURL::Authority{.host = "s3." + region + ".amazonaws.com"};
+        }
+
+        return {httpsUrl.to_string(), parsed};
+    }
+#endif
     void enqueueFileTransfer(const FileTransferRequest & request, Callback<FileTransferResult> callback) override
     {
-        /* Ugly hack to support s3:// URIs. */
+        /* Handle s3:// URIs with curl-based AWS SigV4 authentication */
         if (hasPrefix(request.uri, "s3://")) {
-            // FIXME: do this on a worker thread
-            try {
 #if NIX_WITH_S3_SUPPORT
-                auto parsed = ParsedS3URL::parse(request.uri);
-
-                std::string profile = parsed.profile.value_or("");
-                std::string region = parsed.region.value_or(Aws::Region::US_EAST_1);
-                std::string scheme = parsed.scheme.value_or("");
-                std::string endpoint = parsed.getEncodedEndpoint().value_or("");
-
-                S3Helper s3Helper(profile, region, scheme, endpoint);
-
-                // FIXME: implement ETag
-                auto s3Res = s3Helper.getObject(parsed.bucket, parsed.key);
-                FileTransferResult res;
-                if (!s3Res.data)
-                    throw FileTransferError(NotFound, {}, "S3 object '%s' does not exist", request.uri);
-                res.data = std::move(*s3Res.data);
-                res.urls.push_back(request.uri);
-                callback(std::move(res));
+            // Use curl-based approach with AWS SigV4 authentication
+            enqueueItem(std::make_shared<TransferItem>(*this, request, std::move(callback)));
 #else
-                throw nix::Error("cannot download '%s' because Nix is not built with S3 support", request.uri);
+            throw nix::Error(
+                "cannot download '%s' because Nix is not built with AWS CRT support (requires aws-crt-cpp and curl >= 7.75.0)",
+                request.uri);
```

Comment:
I think this needs to be `callback.throw(...)` or something like that. Janky async.